### PR TITLE
Replace custom memoize module with cachetools

### LINKED ISSUE #4506: Hash collisions in memoization result in invalid values
Our custom `memoize` function is mapping inputs to their hashes, then those hashes are mapped to cached values, and, because the inputs and the hashes' original values aren't compared, a simple hash collision will cause `memoize` to return the wrong values.

Here's a simple example:

```python
import pymc3 as pm


class Bad1:
    def __hash__(self):
        return 329


class Bad2:
    def __hash__(self):
        return 329


b1 = Bad1()
b2 = Bad2()


# Here's what we expect:
d = {}

d[b1] = 1
d[b2] = 2

assert hash(b1) == hash(b2)
assert d[b1] != d[b2]

# Because `memoize` doesn't use the inputs as keys in the `dict`-based
# cache, we don't get the expected post-hash-lookup equality checks that
# prevent errors when faced with hash collisions

@pm.memoize.memoize
def test_func(x):
    return x


assert b1 != b2
assert test_func(b1) != test_func(b2)
```

The last `assert` will currently fail, although it should be equivalent to the `assert` preceding it.

### PR DESCRIPTION
**Pull request by brandonwillard**
This PR replaces our custom `pymc3.memoize` module with `cachetools`.

Closes #4506.

### CODE REVIEW

**Review by twiecki — APPROVED** [review]

**Comment by twiecki** [comment]
Needs a line in release-notes.

**Comment by brandonwillard** [comment]
I changed all the caches to LRUs with a maximum of 128 items.  This should help prevent memory leaks&mdash;at the cost of speed in certain cases, of course.

File: RELEASE-NOTES.md [hunk 1]
```diff
@@ -3,6 +3,7 @@
 ## PyMC3 vNext (TBD)
 ### Breaking Changes
 - ⚠ Theano-PyMC has been replaced with Aesara, so all external references to `theano`, `tt`, and `pymc3.theanof` need to be replaced with `aesara`, `aet`, and `pymc3.aesaraf` (see [4471](https://github.com/pymc-devs/pymc3/pull/4471)).
+- ⚠ The `pymc3.memoize` module was removed and replaced with `cachetools`.  The `hashable` function and `WithMemoization` class were moved to `pymc3.util`.
```
Comment (twiecki) [inline RELEASE-NOTES.md]: This should be under `Maintenance` I think.

File: pymc3/tests/test_util.py [hunk 2]
```diff
@@ -167,3 +169,53 @@ def test_dtype_error(self):
             raise pm.exceptions.DtypeError("With types.", actual=int, expected=str)
         assert "int" in exinfo.value.args[0] and "str" in exinfo.value.args[0]
         pass
+
+
+def test_hashing_of_rv_tuples():
```
Comment (twiecki) [inline pymc3/tests/test_util.py:174]: and this test fails without this change?

Comment (brandonwillard) [reply pymc3/tests/test_util.py:174]: The test `test_hash_key` will.

File: RELEASE-NOTES.md (hunk 1 above)
Comment (brandonwillard) [reply RELEASE-NOTES.md]: OK, I'll change that.

**Review by twiecki — APPROVED** [review]

**Comment by ferrine** [comment]
Thanks, @brandonwillard!

**Pull request merged**